File: js/binary_tree_link/BinaryTreeLink.js
// define node
function BiTNode(props){
    this.data = props.data
    this.lchild = null
    this.rchild = null
}
var MAXSIZE = 100

function BinaryTreeLink(props){}

var btl = new BinaryTreeLink({})

BinaryTreeLink.prototype.visit=function(element){
    process.stdout.write(element + ' ')
}

var strArr = new Array(24)

BinaryTreeLink.prototype.StrAssign = function(T,chars)
{
    if (chars.length>MAXSIZE)
        return false
    else
    {
        T[0] = chars.length
        for (i = 1; i <= T[0]; i++)
            T[i] = chars[i-1]
        return true
    }
}

// Construct empty binary tree
BinaryTreeLink.prototype.InitBiTree = function(T)
{
    // Direct to make Object null, is no use
    T = null
    return true
}

/*
	Condition: binary tree is exist
	Result: destroy binary tree
*/
BinaryTreeLink.prototype.DestroyBiTree = function(T)
{
    if (T)
    {
        // If tree has left child
        if (T.lchild)
            // Destroy subtree of left child
            btl.DestroyBiTree(T.lchild)
        // If tree has right child
        if (T.rchild)
            // Destroy subtree of right child
            btl.DestroyBiTree(T.rchild)
        // Relese root node
        T.data = null
        T.lchild = null
        T.rchild = null
    }
}


index = 1
// "#" means empty tree
BinaryTreeLink.prototype.CreateBiTree = function(T)
{
    var ch
    ch = strArr[index++]

    if (ch == '#')
        T = null
    else
    {
        if (!T)
            return false
        // Generate root node
        T.data = ch
        T.lchild = new BiTNode({})
        T.rchild = new BiTNode({})
        // Construct left subtree
        btl.CreateBiTree(T.lchild)
        // Construct right subtree
        btl.CreateBiTree(T.rchild)
    }
}

/*
	Condition: binary tree is exist
	Result: If tree is empty, return true, otherwise return false
*/
BinaryTreeLink.prototype.BiTreeEmpty = function(T)
{
    if (T.data != undefined)
        return false
    else
        return true
}

/*
	Condition: binary tree is exist
	Result: return the length of tree
*/
BinaryTreeLink.prototype.BiTreeDepth = function(T)
{
    var i, j
    if (!T)
        return 0
    if (T.lchild)
        i = btl.BiTreeDepth(T.lchild)
    else
        i = 0;
    if (T.rchild)
        j = btl.BiTreeDepth(T.rchild)
    else
        j = 0
    return i>j ? i+1 : j+1
}

/*
	Condition: binary tree is exist
	Result: return root of the tree
*/
BinaryTreeLink.prototype.Root = function(T)
{
    if (btl.BiTreeEmpty(T))
        return 0
    else
        return T.data
}

/*
	Condition: binary tree is exist, p points to a node of the tree
	result: return data, which is pointed by p
*/
BinaryTreeLink.prototype.Value = function(p)
{
    return p.data
}

// Assign value to node, which is pointed by p
BinaryTreeLink.prototype.Assign = function(p, value)
{
    p.data = value
}

/*
	Condition: binary tree is exist
	Result: preorder traverse
*/
BinaryTreeLink.prototype.PreOrderTraverse = function(T)
{
    if (T == null)
        return false
    console.log("%s", T.data)
    btl.PreOrderTraverse(T.lchild)
    btl.PreOrderTraverse(T.rchild)
}

/*
	Condition: binary tree is exist
	Result: inorder traverse
*/
BinaryTreeLink.prototype.InOrderTraverse = function(T)
{
    if (T == null)
        return false
    btl.InOrderTraverse(T.lchild)
    console.log("%s", T.data)
    btl.InOrderTraverse(T.rchild)
}

/*
	Condition: binary tree is exist
	Result: postorder traverse
*/
BinaryTreeLink.prototype.PostOrderTraverse = function(T)
{
    if (T == null)
        return false
    btl.PostOrderTraverse(T.lchild)
    btl.PostOrderTraverse(T.rchild)
    console.log("%s", T.data)
}
/*
		 A
	   /   \
	  B      C
	 /  \   /  \
   D    E  F    G
  /       /      \
 H       I        J
  \
   K
*/



var BinarTreeLink = new BiTNode({})
var i;
var T;
var e1;
btl.InitBiTree(BinarTreeLink)


btl.StrAssign(strArr,"ABDH#K###E##CFI###G#J##")

btl.CreateBiTree(BinarTreeLink)

console.log("After constructing empty binary tree, if the tree is empty? %d(1:true 0:false) depth of the tree is=%d\n", btl.BiTreeEmpty(BinarTreeLink), btl.BiTreeDepth(BinarTreeLink))
e1 = btl.Root(BinarTreeLink)
console.log("Root of binary tree is: %c\n", e1)

console.log("\n Preorder traverse:")
btl.PreOrderTraverse(BinarTreeLink)
console.log("\n Inorder traverse:")
btl.InOrderTraverse(BinarTreeLink)
console.log("\n Postorder traverse:")
btl.PostOrderTraverse(BinarTreeLink)
btl.DestroyBiTree(BinarTreeLink)
console.log("\n After clearing binary tree, if the tree is empty? %d(1:true 0:false) depth of the tree is=%d\n", btl.BiTreeEmpty(BinarTreeLink), btl.BiTreeDepth(BinarTreeLink))
i = btl.Root(BinarTreeLink)
if (!i)
    console.log("Tree is empty, no root\n")

File: java/binary_tree_link/src/BinaryTreeLink.java
/**
 * @Author: 谷天乐
 * @Date: 2020/8/21 15:38
 * @Description:
 */
public class BinaryTreeLink {
    public final int MAXSIZE = 100;
    public static String[] strArr = new String[24];

    public boolean visit(String c){
        System.out.print(c+" ");
        return true;
    }

    public boolean StrAssign(String[] T,String chars)
    {
        if (chars.length()>MAXSIZE)
            return false;
        else
        {
            T[0] = String.valueOf(chars.length());
            for (int i = 1; i <= Integer.parseInt(T[0]); i++)
                T[i] = chars.substring(i-1,i);
            return true;
        }
    }

    // Construct empty binary tree
    public boolean InitBiTree(BiTNode T)
    {
        // Direct to make Object null, is no use
        T = null;
        return true;
    }

    /*
        Condition: binary tree is exist
        Result: destroy binary tree
    */
    public void DestroyBiTree(BiTNode T)
    {
        if (T!=null)
        {
            // If tree has left child
            if (T.lchild!=null)
                // Destroy subtree of left child
                this.DestroyBiTree(T.lchild);
            // If tree has right child
            if (T.rchild!=null)
                // Destroy subtree of right child
                this.DestroyBiTree(T.rchild);
            // Relese root node
            T.data = "";
            T.lchild = null;
            T.rchild = null;
        }
    }

    int index = 1;
    // "#" means empty tree
    public boolean CreateBiTree(BiTNode T)
    {
        String ch;
        ch = strArr[index++];

        if (ch.equals("#"))
            T.data="";
        else
        {
            if (T==null)
                return false;
            // Generate root node
            T.data = ch;
            T.lchild = new BiTNode();
            T.rchild = new BiTNode();
            // Construct left subtree
            this.CreateBiTree(T.lchild);
            // Construct right subtree
            this.CreateBiTree(T.rchild);
        }
        return true;
    }

    /*
        Condition: binary tree is exist
        Result: If tree is empty, return true, otherwise return false
    */
    public boolean BiTreeEmpty(BiTNode T)
    {
        if (T!=null)
            return false;
        else
            return true;
    }

    /*
        Condition: binary tree is exist
        Result: return the length of tree
    */
    public int BiTreeDepth(BiTNode T)
    {
        int i, j;
        if (T==null)
            return 0;
        if (T.lchild!=null)
            i = BiTreeDepth(T.lchild);
        else
            i = 0;
        if (T.rchild!=null)
            j = BiTreeDepth(T.rchild);
        else
            j = 0;
        return i>j ? i + 1 : j + 1;
    }

    /*
        Condition: binary tree is exist
        Result: return root of the tree
    */
    public String Root(BiTNode T)
    {
        if (BiTreeEmpty(T))
            return "";
        else
            return T.data;
    }

    /*
        Condition: binary tree is exist, p points to a node of the tree
        result: return data, which is pointed by p
    */
    public String Value(BiTNode p)
    {
        return p.data;
    }

    // Assign value to node, which is pointed by p
    public void Assign(BiTNode p,String value)
    {
        p.data = value;
    }

    /*
        Condition: binary tree is exist
        Result: preorder traverse
    */
    public void PreOrderTraverse(BiTNode T)
    {
        if (T == null)
            return;
        System.out.printf("%s", T.data);
        PreOrderTraverse(T.lchild);
        PreOrderTraverse(T.rchild);
    }

    /*
        Condition: binary tree is exist
        Result: inorder traverse
    */
    public void InOrderTraverse(BiTNode T)
    {
        if (T == null)
            return;
        InOrderTraverse(T.lchild);
        System.out.printf("%s", T.data);
        InOrderTraverse(T.rchild);
    }

    /*
        Condition: binary tree is exist
        Result: postorder traverse
    */
    public void PostOrderTraverse(BiTNode T)
    {
        if (T == null)
            return;
        PostOrderTraverse(T.lchild);
        PostOrderTraverse(T.rchild);
        System.out.printf("%s", T.data);
    }

    /*
             A
           /   \
          B      C
         /  \   /  \
       D    E  F    G
      /       /      \
     H       I        J
      \
       K
    */

    public static void main(String[] args) {
        BiTNode binarTreeLink = new BiTNode();
        BinaryTreeLink btl = new BinaryTreeLink();
        btl.InitBiTree(binarTreeLink);

        btl.StrAssign(strArr, "ABDH#K###E##CFI###G#J##");

        btl.CreateBiTree(binarTreeLink);

        System.out.printf("After constructing empty binary tree, if the tree is empty? %b(1:true 0:false) depth of the tree is=%d\n", btl.BiTreeEmpty(binarTreeLink), btl.BiTreeDepth(binarTreeLink));
        String e1 = btl.Root(binarTreeLink);
        System.out.printf("Root of binary tree is: %s\n", e1);

        System.out.printf("\n Preorder traverse:");
        btl.PreOrderTraverse(binarTreeLink);
        System.out.printf("\n Inorder traverse:");
        btl.InOrderTraverse(binarTreeLink);
        System.out.printf("\n Postorder traverse:");
        btl.PostOrderTraverse(binarTreeLink);
        btl.DestroyBiTree(binarTreeLink);
        System.out.printf("\n After clearing binary tree, if the tree is empty? %b(1:true 0:false) depth of the tree is=%d\n", btl.BiTreeEmpty(binarTreeLink), btl.BiTreeDepth(binarTreeLink));
        String i = btl.Root(binarTreeLink);
        if (i=="")
            System.out.printf("Tree is empty, no root\n");
    }

}